build-logic/src/main/java/com/squareup/workflow1/buildsrc/shardConnectedChecks.kt

```diff
@@ -0,0 +1,111 @@
+ * the `connectedCheckShardMatrixYamlUpdate` task can automatically update the workflow file so
+ * that they're all invoked.
+ *
+ * The shard tasks are invoked as:
+ * ```shell
+ * # roughly 1/3 of the tests
+ * ./gradlew connectedCheckShard1
+ * # the second third
+ * ./gradlew connectedCheckShard2
+ * # the last third
+ * ./gradlew connectedCheckShard3
+ * ```
+ *
+ * @param target the root project which gets the shard tasks
+ * @param shardCount the number of shard tasks to create.
+ */
+fun shardConnectedCheckTasks(
+  target: Project,
+  shardCount: Int = 3
+) {
+  if (target != target.rootProject) {
+    throw GradleException("Only add connectedCheck shard tasks from the root project.")
+  }
+
+  target.registerYamlShardsTasks(
+    shardCount = shardCount,
+    startTagName = "### <start-connected-check-shards>",
+    endTagName = "### <end-connected-check-shards>",
+    taskNamePart = "connectedCheck",
+    yamlFile = target.rootProject.file(".github/workflows/kotlin.yml")
+  )
+
+  // Assign each project to a shard.
+  // It's lazy so that the work only happens at task configuration time, but it's outside the
+  // task configuration block so that it only happens once.
+  val shardAssignments by lazy {
+
+    val testAnnotationRegex = "@Test(?!Factory)".toRegex()
+    val testFactoryAnnotationRegex = "@TestFactory".toRegex()
+
+    // Calculate the cost of each project's tests
+    val projectTestCosts = target.subprojects
+      // Only Android projects can have these tasks.
+      // Use the KGP Android plugin instead of AGP since KGP has only one ID to look for.
+      .filter { it.plugins.hasPlugin("org.jetbrains.kotlin.android") }
+      // If the project doesn't have any test sources, don't bother adding it to a shard.
+      .filter { it.file("src/androidTest").exists() }
+      .associateWith { project ->
+        project.file("src/androidTest")
+          .walkTopDown()
+          .filter { it.isFile && it.extension == "kt" }
+          .sumOf { file ->
+            val fileText = file.readText()
+            val testAnnotationCount = testAnnotationRegex.findAll(fileText).count()
+            val testFactoryAnnotationCount =
+              testFactoryAnnotationRegex.findAll(fileText).count()
+            testAnnotationCount + (testFactoryAnnotationCount * 2)
+          }
+      }
+
+    // Sort the projects by descending test cost, then fall back to the project paths
+    // The path sort is just so that the shard composition is stable.  If the shard composition
+    // isn't stable, the shard tasks may not be up-to-date and build caching in CI is broken.
+    val sortedProjects = projectTestCosts.keys
+      .sortedWith(
+        compareBy(
+          { projectTestCosts.getValue(it) },
+          { it.path }
+        )
+      )
+
+    var shardIndex = 0
+
+    sortedProjects.groupBy { (shardIndex++ % shardCount) + 1 }
+  }
+
+  (1..shardCount).map { shardIndex ->
```

Comment:
If we have an off by one error and some boundary tests fall between the cracks, how will we notice that shard1 + shard2 + shard3 < allShards? Wondering if there's an invariant check we can put somewhere. 🤔

Reply:
Maybe extract a core of this routine so that you can unit test it?

Reply:
I didn't (and still don't) want to introduce unit tests to `build-logic`, since the tasks for an included build are not hooked into Gradle's implicit task delegation by default.

For example, if we invoke `./gradlew test` today, it'll invoke all unit tests in kotlin-jvm modules.  We just expect that to work.  But that hook (implicit task delegation) only applies to sub-projects. The `build-logic` build is not affected by this, so its `test` task would never be invoked.

We can manually stitch things together, such as with this in the root `build.gradle.kts`:

```kotlin
tasks.named("test") {
  dependsOn(gradle.includedBuild("build-logic").task(":test"))
}
```

...but if anything were to happen to that, then the sharding could break in the way you described and no one would be the wiser.

So, instead I refactored it to be as off-by-one-proof as I could.  As the tasks are configuring, they assert that the shard allocations total up to the exact number of Android projects and there are no duplicates.  So they test themselves every time they're used.